packages/dd-trace/src/debugger/devtools_client/snapshot/collector.js

```diff
@@ -14,56 +15,66 @@ module.exports = {
 // each lookup will just finish in its own time and traverse the child nodes when the event loop allows it.
 // Alternatively, use `Promise.all` or something like that, but the code would probably be more complex.
 
-async function getObject (objectId, maxDepth, depth = 0) {
-  const { result, privateProperties } = await session.post('Runtime.getProperties', {
+async function getObject (objectId, opts, depth = 0, collection = false) {
+  let { result, privateProperties } = await session.post('Runtime.getProperties', {
     objectId,
     ownProperties: true // exclude inherited properties
   })
 
-  if (privateProperties) result.push(...privateProperties)
+  if (collection) {
+    // Trim the result if it's too large.
+    // Collections doesn't contain private properties, so the code in this block doesn't have to deal with it.
+    const size = countEnumerableProperties(result)
+    if (size > opts.maxCollectionSize) {
+      result = result.slice(0, opts.maxCollectionSize)
+      result[collectionSizeSym] = size
+    }
+  } else if (privateProperties) {
+    result.push(...privateProperties)
+  }
```

Comment:
Note: It's on purpose that I do not take into account the scenario where `collection === true` and `privateProperties !== undefined`. This is because, currently, the only time `collection === true`, is when we're either getting the properties of an `Array` or a `TypedArray`. For those two object types, we know in advance that there's no private properties, so we can make the code simpler by not having to deal with that edge case.

If we wanted to deal with this, the code would look like this:

```suggestion
  if (collection) {
    const privatePropertiesSize = privateProperties?.length ?? 0
    const totalSize = countEnumerableProperties(result) + privatePropertiesSize

    if (totalSize > opts.maxCollectionSize) {
      // Prioritize private properties over regular properties
      const spaceForNonPrivateProperties = Math.max(opts.maxCollectionSize - privatePropertiesSize, 0)
      result = result.slice(0, spaceForNonPrivateProperties)
      result[collectionSizeSym] = totalSize

      if (privateProperties) {
        const remainingSize = opts.maxCollectionSize - result.length
        if (privatePropertiesSize > remainingSize) {
          privateProperties = privateProperties.slice(0, remainingSize)
        }
      }
    }
  }

  if (privateProperties) result.push(...privateProperties)
```